packages/react-pages/src/client/state.ts

```diff
@@ -113,7 +113,11 @@ if (import.meta.hot) {
 
   usePageModule = (pagePath) => {
     const data = useAtomValue(dataAtoms(pagePath))
-    return useMemo(() => data?.data(), [data])
+    // if the dataModulePath hasn't changed,
+    // we can keep using the same dataModule.
+    // vite hmr will handle it's update.
+    // (all dataModules should be self-accepted)
+    return useMemo(() => data?.data(), [data.dataModulePath])
```

Comment:
This doesn't fix the problem entirely. The caller of `usePageModule` will still re-render.

Jotai atoms must return a stable object reference or a primitive to avoid accidental re-rendering. I suggest changing `dataAtoms` back to `dataPathAtoms`.

Reply:
If we change back to `dataPathAtoms`, how can we avoid ` import(dataPath)` in this code:
```ts
  const dataPathAtoms = atomFamily((path: string) => (get) => {
    const pages = get(pagesAtom)
    const page = pages[path] || pages['/404']
    return page?.dataPath || null
  })

  usePageModule = (pagePath) => {
    const dataPath = useAtomValue(dataPathAtoms(pagePath))
    return useMemo(() => {
      return dataPath ? import(dataPath /* @vite-ignore */) : void 0
    }, [dataPath])
  }
```

Reply:
Are you sure there was even a problem to begin with? Please give the reproduction steps to prove the bug exists in the current implementation (before this PR). I'm using the current implementation without experiencing full reloads on page changes.

Reply:
checkout this branch: https://github.com/vitejs/vite-plugin-react-pages/tree/hmr-bug
git clean -qfdx
pnpm i
cd packages/playground/custom-find-pages
npm run dev
visit http://localhost:3000/Button
delete file `packages/playground/custom-find-pages/src/Button/demos/demo2.tsx`
full reload will happen

Reply:
I don't think there is an easy way to handle the hmr of module such as `/@react-pages/pages/page1`, which is something like:
```js
import * as m1 from "/runtime-module-path1";
import * as m2 from "/runtime-module-path2";

const modules = {};
modules["key1"] = m1;
modules["key2"] = m2;

export default modules;
```

So I am going to stop trying to do the most ideal hmr updating(preventing all unnecessary re-render).